services/user/XAdmin/ui/src/components/forms/key-device-form.tsx

```diff
@@ -0,0 +1,158 @@
+import { UseFormReturn } from "react-hook-form";
+import { z } from "zod";
+
+import { KeyDeviceSchema } from "@/pages/create-page";
+import { getErrorMessage } from "@/lib/utils";
+
+import {
+    Form,
+    FormControl,
+    FormField,
+    FormItem,
+    FormLabel,
+    FormMessage,
+} from "../ui/form";
+import {
+    Card,
+    CardContent,
+    CardDescription,
+    CardHeader,
+    CardTitle,
+} from "../ui/card";
+import { Input } from "../ui/input";
+import { RadioGroup, RadioGroupItem } from "../ui/radio-group";
+
+import { useKeyDevices, useUnlockKeyDevice } from "../../hooks/useKeyDevices";
+
+interface Props {
+    form: UseFormReturn<z.infer<typeof KeyDeviceSchema>>;
+    next: () => Promise<void>;
+}
+
+export const KeyDeviceForm = ({ form, next }: Props) => {
+    const { data: keyDevices, isError, isSuccess } = useKeyDevices();
+    const { mutateAsync: unlock } = useUnlockKeyDevice();
+
+    const keyDeviceId = form.watch("id");
+    const selectedDevice = keyDevices?.find((d) => d.id === keyDeviceId);
+
+    const unlockDevice = async (pin?: string) => {
+        try {
+            await unlock({ device: keyDeviceId, pin });
+            return true;
+        } catch (e) {
+            const message = getErrorMessage(e);
+            if (message === "pin incorrect") return "Incorrect PIN";
+            return message;
+        }
+    };
+
+    const onSubmit = (e: React.SyntheticEvent) => {
+        e.preventDefault();
+        next();
+    };
+
+    return (
+        <Card className="min-w-[350px]">
+            <CardHeader>
+                <CardTitle>Select key device</CardTitle>
+                <CardDescription>
+                    Where do you want your block producer server key to be
+                    stored?
+                </CardDescription>
+            </CardHeader>
+            <CardContent>
+                {isSuccess && !keyDevices.length ? (
+                    <div className="mb-4 flex max-w-[450px] items-center gap-3">
+                        <p className="text-2xl">⚠️</p>
+                        <p>
+                            No key devices were found. Please ensure one is
+                            available. If you'd like to boot without a key
+                            device, go back and select the "Development" option.
+                        </p>
+                    </div>
+                ) : null}
+                {isError ? (
+                    <div className="mb-4 flex max-w-[450px] items-center gap-3">
+                        <p className="text-2xl">⚠️</p>
+                        <p>
+                            There was an error fetching key devices. See the
+                            console for more information.
+                        </p>
+                    </div>
+                ) : null}
```

Comment:
Can we just display the error message here? If it's a simple error the user might be able to recover themselves rather than reach out to a dev.

Reply:
Yeah, I can do that. I figure anyone booting a chain should know how to look at the console. But I don't feel strongly here either way.